market_maker_keeper/uniswapv2_market_maker_keeper.py

```diff
@@ -0,0 +1,379 @@
+                token_b_removed = token_b_balance_after_remove - token_b_balance
+                self.logger.info(f"Real {self.token_a.name} amount removed: {token_a_removed} "
+                                    f"Real {self.token_b.name} amount removed: {token_b_removed} "
+                                    f"tx fee used {tx_fee} "
+                                    f"with tx hash {transact.transaction_hash.hex()}")
+
+                return transact
+            else:
+                self.logger.warning(f"Failed to remove {liquidity_to_remove} liquidity of {self.uniswap.pair_address.address}")
+        else:
+            self.logger.info(f"No liquidity to remove")
+
+    def determine_liquidity_action(self) -> Tuple:
+        """
+        Add or remove liquidity depending upon the difference between Uniswap ratio and our external price feeds.
+
+        Pricing is expressed as a function of token_a balance / token_b balance in the contract
+
+        First calculate the acceptable price movement limit based upon the
+        difference between external price feeds and the accepted slippage.
+
+        If the accepted difference is greater than the distance of uniswaps price from external prices
+        add liquidity to the pool, otherwise remove it.
+        """
+
+        self.feed_price = self.price_feed.get_price().buy_price if self.testing_feed_price is False else self.test_price
+
+        self.logger.info(f"Feed price: {self.feed_price} Uniswap price: {self.uniswap_current_exchange_price}")
+
+        if self.feed_price is None:
+            self.logger.warning(f"Price feed is returning null, removing all available liquidity")
+            add_liquidity = False
+            remove_liquidity = True
+        else:
+            diff = self.feed_price * self.accepted_slippage
+
+            add_liquidity = diff > abs(self.feed_price - self.uniswap_current_exchange_price)
+            remove_liquidity = diff < abs(self.feed_price - self.uniswap_current_exchange_price)
+
+            self.logger.info(
+                f"Feed price / Uniswap price diff {diff} triggered add liquidity: {add_liquidity}; remove liquidity: {remove_liquidity}")
+
+        return add_liquidity, remove_liquidity
+
+    def place_liquidity(self):
+        """
+
+        """
+
+        exchange_token_a_balance = Wad.from_number(0) if self.uniswap.is_new_pool else self.uniswap.get_exchange_balance(self.token_a, self.uniswap.pair_address)
+        exchange_token_b_balance = Wad.from_number(0) if self.uniswap.is_new_pool else self.uniswap.get_exchange_balance(self.token_b, self.uniswap.pair_address)
+
+        self.logger.info(f"Exchange Contract {self.token_a.name} amount: {exchange_token_a_balance}; "
+                         f"Exchange Contract {self.token_b.name} amount: {exchange_token_b_balance}")
+
+        add_liquidity, remove_liquidity = self.determine_liquidity_action()
```

Comment:
We can add an assert, but I don't think this is necessary - As written it's not currently logically possible for them to both be true simultaneously. `diff` would have to be simultaneously be both smaller and larger than `feed_price - uniswap_current_exchange_price`. If the underlying numbers are None, it will trigger the `if case` and `remove_liquidity` would be hardcoded `True`, and `add_liquidity` would be `False`.

Reply:
Given simplicity and minimal cost, i've added an assert to check that the values aren't equal.

Reply:
Thank you, always best just in case. Yes, I agree with you I don't see the logical possibility but best to protect us from those 
crazy scenarios. Just my opinion.